library/alloc/src/macros.rs

```diff
@@ -47,10 +47,14 @@ macro_rules! vec {
         $crate::vec::from_elem($elem, $n)
     );
     ($($x:expr),+ $(,)?) => (
-        <[_]>::into_vec(
-            // Using the intrinsic produces a dramatic improvement in stack usage for
-            // unoptimized programs using this code path to construct large Vecs.
-            $crate::boxed::box_new([$($x),+])
+        // Using `write_box_via_move` produces a dramatic improvement in stack usage for unoptimized
+        // programs using this code path to construct large Vecs. We can't use `write_via_move`
+        // because this entire invocation has to remain a call chain without `let` bindings, or else
+        // the temporary scopes change and things break (see `vec-macro-rvalue-scope` test).
+        // This function isn't actually safe but the way we use it here is. We can't use an
+        // unsafe block as that would also wrap `$x`.
+        $crate::boxed::box_uninit_array_into_vec_unsafe(
+            $crate::intrinsics::write_box_via_move($crate::boxed::Box::new_uninit(), [$($x),+])
         )
```

Comment:
Has an immediately invoked closure been considered?
(Though having an added closure for each occurance of vec! might have problems on its own. Is there a way to mark the closure inline(always) to prevent extra function calls when using a debug build?)

```suggestion
(|b| unsafe{
    $crate::boxed::box_uninit_array_into_vec_unsafe(b)
})(
    $crate::intrinsics::write_box_via_move(
        $crate::boxed::Box::new_uninit(), [$($x),+]
    )
)
```